### ISSUE TITLE
Add tracking for the various block and pattern inserters in Gutenberg

### ISSUE THREAD

**Issue by simison** [Automattic/wp-calypso#50426]
We track block & pattern insertions in .com block editor, but we don't have visibility _which_ inserter folks use.

We need to know separately if a block was inserted via:

• top inserter
• quick inserter
• sibling inserter (between blocks)
• inline inserter (slash command)

<img width="1033" alt="Screenshot 2021-02-24 at 13 56 54" src="https://user-images.githubusercontent.com/87168/108997625-7d145580-76a8-11eb-9b05-45b5792ba359.png">

There might be some separation of tracks events already when searching for blocks, see:

> https://github.com/Automattic/wp-calypso/blob/master/apps/wpcom-block-editor/src/wpcom/features/tracking/wpcom-inserter-menu-search-term.js#L33
> https://github.com/Automattic/wp-calypso/blob/master/apps/wpcom-block-editor/src/wpcom/features/tracking/wpcom-inserter-inline-search-term.js#L34
>
> That gets tracked in conjunction with search, so it might not be part of blocks that are selected from the inserter without a search
>There’s a context attribute that gets tracked with block insertions
> it only diffenciates between / command inline and the block selector as a whole

via Slack convo p1605037273252600-gutenberg-slack

Older related issue https://github.com/Automattic/wp-calypso/issues/33680

**Comment by simison** [comment]
@autumnfjeld we talked that tracking this could be useful for future editor onboarding initiatives. Could Ganon find time to add these when convenient for you? It's not urgent.

**Comment by simison** [comment]
Some more context for having this data pcjTuq-a3-p2/#comment-318

**Comment by autumnfjeld** [comment]
@simison Ganon discussed this in team meeting today, no problem to add to our "Up Next" after higher priority items.  This also dovetails well with the analytics work I need to do on Welcome Tour & Editor interaction. Helpful Tracking reference: PCYsg-nrf-p2.

**Comment by autumnfjeld** [comment]
Ganon estimates this as Size S (to us this means < week , well defined, and no core work).  We'll open another ticket if core work is needed. 😃

**Comment by p-jackson** [comment]
I've got a WIP for this over at #51531. It's a mixture of relying on redux and DOM state, so I'll need to add some e2e tests so we get notified when this breaks.

There are two existing tracks events
- `wpcom_block_inserted`, when a block is inserted
- `wpcom_block_picker_block_inserted`, when a block is replaced

and I'm adding a new `insert_method` prop to these events that can be either:
- `header-inserter`
- `slash-inserter`
- `quick-inserter`
- `block-switcher`
- or undefined

It will be undefined if the block was inserted in a way that I haven't tracked (I assume we don't need to know every single way). This is things like using WPCOM's page layout picker modal or pressing "return" at the end of a paragraph block. We'll always have to support some sort of "unknown" type because it could be a 3rd party plugin adding blocks in an unknown way.

It will be `block-switcher` if the "change block type or style" button was clicked on the floating block toolbar.

It will be `quick-inserter` if it was any of these 3 style of inserter buttons.
<img width="556" alt="Screenshot 2021-03-30 at 5 45 26 PM" src="https://user-images.githubusercontent.com/1500769/112935178-c102ea80-917f-11eb-995a-c95ecc9277f9.png">
They're all the same component under the hood which is why they're being tracked the same. @annezazu @simison I assume we don't need to tell these 3 types of inserters apart? From pcjTuq-a3-p2/#comment-318 I think we mainly just want to know whether `header-inserter` or `quick-inserter` will be more popular?

**Comment by simison** [comment]
Thanks @p-jackson ! Sounds good — yeah I'm personally interested in header vs quick-inserter mostly myself.

> It will be quick-inserter if it was any of these 3 style of inserter buttons.

Do inserters inside Buttons block and Social icons block get tracked in this group as well? I don't remember if FSE Navigation block had any of these inserters but if it does, same question. :-)

<img width="225" alt="Screenshot 2021-03-30 at 14 03 02" src="https://user-images.githubusercontent.com/87168/112979076-c69d0800-9160-11eb-81f4-7b362c367f54.png">
<img width="231" alt="Screenshot 2021-03-30 at 14 02 03" src="https://user-images.githubusercontent.com/87168/112979081-c7359e80-9160-11eb-94a6-12b098beb874.png">

cc @mtias and @apeatling in case you spot anything missing or had thoughts on tracking granularity.

**Comment by annezazu** [comment]
> They're all the same component under the hood which is why they're being tracked the same. 

I think that's fine for the screenshot you shared.

>  think we mainly just want to know whether header-inserter or quick-inserter will be more popular?

I actually am curious more so about the various ways people insert things including the other methods you listed like slash inserter rather than just those two! All of them provide insights into the common pathways people interact with blocks.

**Comment by p-jackson** [comment]
The just deployed the new `insert_method` prop D60140-code.
Needs tests because I think this will inevitably break in a future version of GB, but it's going to be easier to write those e2e tests now that the new version of `wpcom-block-editor` is in production. Without that it was hard to get tests pointing at the changes in my sandbox.

**Comment by autumnfjeld** [comment]
@simison @p-jackson Can we close this issue since work in https://github.com/Automattic/wp-calypso/pull/51531 is merged?

**Comment by simison** [comment]
@autumnfjeld yep! I'll close this.

There are some follow-up questions in the post (pbAok1-28a-p2), these might need their own issues if they can be improved;

- What is the difference between `null` and `undefined` values?
- Can we capture when blocks are inserted from page template picker so that those wouldn't appear as undefined?

**Comment by p-jackson** [comment]
> What is the difference between null and undefined values?

I don't know where both the `null` and `undefined` values are coming from! It should only be `undefined`. Maybe Tracks is doing something special with nullish prop values? Maybe we should use `'unknown'` instead.

> Can we capture when blocks are inserted from page template picker so that those wouldn't appear as undefined?

Yep doable. Very wpcom specific, but I think [there's a filter that's set to `true` when the modal is adding blocks](https://github.com/Automattic/wp-calypso/blob/4d8155849c9a164f1b964adcd4528f149399259d/apps/editing-toolkit/editing-toolkit-plugin/starter-page-templates/page-patterns-plugin.tsx#L63) that we can use to detect whether the modal is the thing doing the inserting.

**Comment by simison** [comment]
> Maybe Tracks is doing something special with nullish prop values? Maybe we should use 'unknown' instead.

I like your idea to explicitly setting _unknown_ also because it allows us to detect if null/undefined is coming from outside. I think `null` appears when the attribute is completely missing, for example, but not sure!

You didn't suggest "other" as a value but I wanted to mention specifically that "unknown" is also semantically better than "other" because we don't know which all future inserters would fall into that, so it truly is _unknown_.

> Very wpcom specific, 

Anything I'm missing, or aren't Tracks events always wpcom specific? :-)

**Comment by simison** [comment]
>  I think there's a filter that's set to true when the modal is adding blocks that we can use to detect whether the modal is the thing doing the inserting.

 👍 This would be a great addition, can you open up an issue for the Ganon board? Thx! 

Do blocks inserted from pattern inserter also appear as undefined/null?

**Comment by p-jackson** [comment]
> Can we capture when blocks are inserted from page template picker so that those wouldn't appear as undefined?

> Anything I'm missing, or aren't Tracks events always wpcom specific? :-)

I think I feel victim to my own fallacious logic somewhere in there, but I'm not sure where 😅 So scratch all that!
Yes, we should be able to track block insertions that come from the page layout picker modal. I'll add it to the Ganon board. Along with a task to change `undefined` values to `"unknown"`.

> Do blocks inserted from pattern inserter also appear as undefined/null?

No they're marked as coming from the inserter that was used to insert the pattern. Same as reusable blocks (which are insertable with the slash inserter)

**Comment by annezazu** [comment]
Noting that we still show null in the data (and it's by far the largest grouping).

**Comment by p-jackson** [comment]
I tested in the page editor and the event appeared to be working as I expected it to.
But when I tested in the site editor there were a lot of `wpcom_block_inserted` and `wpcom_block_picker_block_inserted` events being fired. Even just navigating around using the sidebar would fire these events! And they're missing the `insert_method` prop, which is probably what's causing all the `null` values in the data.

My guess is we simply never updated these tracks events to work correctly with the site editor.

Correction: there was a lot of work done https://github.com/Automattic/wp-calypso/issues/53410
